src/ProcessEpoch.hpp

```diff
@@ -0,0 +1,106 @@
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef PROCESSEPOCH_HPP_INCLUDE
+#define PROCESSEPOCH_HPP_INCLUDE
+
+#include <cstdint>
+#include <memory>
+
+#include "ApplicationSampler.hpp"
+
+
+namespace geopm
+{
+    /// @brief Class that analyzes application records sent from a
+    ///        single process to determine epoch related signals for
+    ///        the EpochIOGroup.  These signals are:
+    ///            {"EPOCH_RUNTIME",
+    ///             "EPOCH_COUNT",
+    ///             "EPOCH_RUNTIME_NETWORK",
+    ///             "EPOCH_RUNTIME_IGNORE"}.
+    class ProcessEpoch
+    {
+        public:
+            /// @brief Make a ProcessEpoch object that will do an
+            ///        analysis of the application sampler record
+            ///        updates.
+            ///
+            ///  Factory method to create a unique pointer to a
+            ///  ProcessEpoch object that uses the M_EVENT_EPOCH_COUNT
+            ///  and M_EVENT_HINT events to determine the epoch
+            ///  related signals.
+            ///
+            ///  @return Unique pointer to a ProcessEpoch base class
+            ///          object that processes epoch events.
+            static std::unique_ptr<ProcessEpoch> make_unique(void);
+            /// @brief Base class constructor method for pure virtual
+            ///        interface.
+            ProcessEpoch() = default;
+            /// @brief Base class destructor method for pure virtual
+            ///        interface.
+            virtual ~ProcessEpoch() = default;
+            /// @brief Process application sampler record to update
+            ///        signal data.
+            ///
+            /// @param record A reference to a record that was queued
+            ///        by the process tracked by this object.  It is
+            ///        the callers responsiblity to filter out records
+            ///        that are sent from other processes.
+            virtual void update(const ApplicationSampler::m_record_s &record) = 0;
+            /// @brief The number of epoch events that have occured
+            ///        for the process that is tracked by this object.
+            ///
+            /// @return Zero based counter of the number of epoch
+            ///         events.
+            virtual int epoch_count(void) const = 0;
+            /// @brief The total runtime that elapsed between the last
```

Comment:
If this is supposed to return NAN before the first epoch completes, we should mention that here